### PR TITLE
Fix cryptographically insecure ID generation

### PR DESCRIPTION
**Pull request by tomauth0**
### Description

The existing methods of ID generation are not cryptographically secure.  This PR introduces the id-generation lib - a lightweight lib for generating suitably random IDs with a given dictionary. [Utils.js updated to make use of this lib](https://github.com/auth0/node-samlp/pull/116/files#diff-5dfe38baf287dcf756a17c2dd63483781b53bf4b669e10efdd01e74bcd8e780aR142)

Also used the opportunity to cleanup some of the utils file - [swapping out vars and removing some unneeded syntax](https://github.com/auth0/node-samlp/pull/116/files#diff-5dfe38baf287dcf756a17c2dd63483781b53bf4b669e10efdd01e74bcd8e780a)

### References

https://github.com/auth0/node-samlp/issues/111

### Testing

Unit tests added for ID length and dictionary memebership

- [x] This change adds test coverage for new/changed/fixed functionality

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR or in auth0.com/docs
- [x] All active GitHub checks for tests, formatting, and security are passing
- [x] The correct base branch is being used, if not `master`

### CODE REVIEW

File: lib/utils.js [hunk 1]
```diff
@@ -85,11 +88,11 @@ module.exports.parseSamlRequest = function(req, samlRequest, type, options, call
...
 
   try {
-    var input = new Buffer(samlRequest, 'base64');
+    const input = new Buffer(samlRequest, 'base64');
 
     if (input[0] === 60) {  // open tag
       // content is just encoded, not zipped
-      var xml = new xmldom.DOMParser().parseFromString(input.toString());
+      const xml = new xmldom.DOMParser().parseFromString(input.toString());
       if (!xml || !xml.documentElement) {
```
Comment (jakelacey2012) [inline lib/utils.js:96]: It looks like this bit of code is the same in the else statement the only difference is that we're unzipping it, I wonder if this bit of code could be abstracted into a function which is passed the result of `.toString()`.
```javascript
const xml = new xmldom.DOMParser().parseFromString(input);

if (!xml || !xml.documentElement) {
   return callback(new Error('Invalid SAML Request'));
}

configureSigningCredentials(xml, function (err, shouldValidate) {
   if (!shouldValidate) {
          return callback(null, xml);
   }
   checkSignature(xml, callback);
});
```
I think doing this should mean we can avoid setting `xml = null` in the next block as well.

**Review by jakelacey2012 — COMMENTED** [review]
I thought I'd have a go at reviewing some code today, let me know what you think of my comments. I don't think I'm comfortable yet in hitting the approve button, so might be best to get another pair of eyes as well :)

File: test/utils.tests.js [hunk 2]
```diff
@@ -11,4 +11,14 @@ describe('utils', function () {
       });
     });
   });
+  describe('generateUniqueID', function() {
+    it('should generate an ID 20 chars long', function() {
+      expect(utils.generateUniqueID().length).to.equal(20);
+    });
+  });
+  describe('generateUniqueID', function() {
+    it('should generate an ID from the alphabet', function() {
```
[the remark this answers is not in the document]
Comment (tomauth0) [reply test/utils.tests.js:20]: So, in this case the 'alphabet' refers to an 'alphabet' parameter on the ID Gen library, as opposed to the english alphabet - it tests that the characters in the resulting ID intersect with the chars passed in the alphabet param on IdGenerator creation

File: lib/utils.js (hunk 1 above)
Comment (luuuis) [reply lib/utils.js:96]: Well spotted. I usually "lift" the optional computation into a local const to keep the computation linear.

```
const input = new Buffer(samlRequest, 'base64');
const inflateIfNeeded = input[0] === 60 ? (cb) => setImmediate(cb, null, input) : zlib.inflateRaw;

inflateIfNeeded(input, function(err, inflated) {
  // shared logic
});
```

**Review by luuuis — APPROVED** [review]

File: lib/utils.js (hunk 1 above)
Comment (jakelacey2012) [reply lib/utils.js:96]: Thats nice, I think I prefer that way!

**Review by ziluvatar — APPROVED** [review]

File: lib/utils.js (hunk 1 above)
Comment (tomauth0) [reply lib/utils.js:96]: This is a great spot and change, but I'm apprehensive to apply it in this particular PR & ticket.
There's a few different areas in the utils file (and across the lib) that could do with a refactor (misspelled opts, etc.).  I'll place a ticket on the backlog that references this suggestion and a few other proposals.

Comment (jakelacey2012) [reply lib/utils.js:96]: Yeah totally, that sounds sensible to me.

**Pull request merged**